core/Symbols.cc

```diff
@@ -579,6 +579,60 @@ bool ClassOrModuleRef::isPackageSpecSymbol(const GlobalState &gs) const {
     return false;
 }
 
+bool ClassOrModuleRef::isBuiltinGenericForwarder() const {
+    return *this == Symbols::T_Hash() || *this == Symbols::T_Array() || *this == Symbols::T_Set() ||
+           *this == Symbols::T_Range() || *this == Symbols::T_Enumerable() || *this == Symbols::T_Enumerator() ||
+           *this == Symbols::T_Enumerator_Lazy();
+}
+
+ClassOrModuleRef ClassOrModuleRef::maybeUnwrapBuiltinGenericForwarder() const {
+    if (*this == Symbols::T_Array()) {
+        return Symbols::Array();
+    } else if (*this == Symbols::T_Hash()) {
+        return Symbols::Hash();
+    } else if (*this == Symbols::T_Enumerable()) {
+        return Symbols::Enumerable();
+    } else if (*this == Symbols::T_Enumerator()) {
+        return Symbols::Enumerator();
+    } else if (*this == Symbols::T_Enumerator_Lazy()) {
+        return Symbols::Enumerator_Lazy();
+    } else if (*this == Symbols::T_Range()) {
+        return Symbols::Range();
+    } else if (*this == Symbols::T_Set()) {
+        return Symbols::Set();
+    } else {
+        return *this;
+    }
+}
+
+ClassOrModuleRef ClassOrModuleRef::forwarderForBuiltinGeneric() const {
+    if (*this == Symbols::Array()) {
+        return Symbols::T_Array();
+    } else if (*this == Symbols::Hash()) {
+        return Symbols::T_Hash();
+    } else if (*this == Symbols::Enumerable()) {
+        return Symbols::T_Enumerable();
+    } else if (*this == Symbols::Enumerator()) {
+        return Symbols::T_Enumerator();
+    } else if (*this == Symbols::Enumerator_Lazy()) {
+        return Symbols::T_Enumerator_Lazy();
+    } else if (*this == Symbols::Range()) {
+        return Symbols::T_Range();
+    } else if (*this == Symbols::Set()) {
+        return Symbols::T_Set();
+    } else {
+        return Symbols::noClassOrModule();
+    }
+}
```

Comment:
This is a nice cleanup.

WDYT about factoring out a constexpr array of stdlib, generic pairs, then updating both of these functions to be basically calls to `std::find_if`, so that it is impossible to forget to update one without the other?  (Ideally the compiler will unroll the loops so that the generated code is equivalent...bonus points for making the stdlib symbols and the generic symbols consecutive so that the functions boil down to just a range check rather than an if-chain.)

I guess once a table is made, the predicate functions before and after this block could just become table lookups too, furthering the goal of not missing places to update if need be?

Reply:
I'm not sure. I'm getting this error from clang:

```plain
clang: Constexpr variable cannot have non-literal type 'const vector<pair<core::ClassOrModuleRef, core::ClassOrModuleRef>>'

vector:471: 'vector<std::pair<sorbet::core::ClassOrModuleRef, sorbet::core::ClassOrModuleRef>>' is not literal because it is not
an aggregate and has no constexpr constructors other than copy or move constructors
```

I'm happy to take this as a follow up, but I'm not quite sure how to solve this. In the mean time, I think the current organization is better than having it scattered throughout the codebase, and I've already got enough stacked PRs at the moment. So I'll merge for now and follow up later.

Reply:
Yeah, we are not on the version of C++ that permits constexpr dynamic memory allocation (:exploding_head:).  Using regular C arrays should work just fine, though.